File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        TreeNode* p_tracker = root;
        TreeNode* q_tracker = root;
        TreeNode* prev;
        while(p_tracker -> val == q_tracker -> val){
            prev = p_tracker;
            if(p -> val < p_tracker -> val)
                p_tracker = p_tracker -> left;
            else if(p -> val > p_tracker -> val) 
                p_tracker = p_tracker -> right;
            else 
                return p_tracker;

            if(q -> val < q_tracker -> val)
                q_tracker = q_tracker -> left;
            else if(q -> val > q_tracker -> val)
                q_tracker = q_tracker -> right;
            else 
                return q_tracker;

        }

        return prev;
    }
};

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        if root is None : return "N"
        return str(root.val) + '.' + self.serialize(root.left) +'.'+ self.serialize(root.right)


        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        elements = data.split('.')

        def build():
            if elements[0] == 'N':
                elements.pop(0)
                return None
            else:
                curr = TreeNode(int(elements[0]))
                elements.pop(0)
                curr.left = build()
                curr.right = build()
                return curr
        
        return build()
        
        
